### PR TITLE
Pod anti affinity admission controller

### LINKED ISSUE #19914: Admission controller to restrict hard ("requiredDuringScheduling") anti-affinity to "node" TopologyKey
As discussed in the (anti-)affinity issue (#18265), "hard' anti-affinity is dangerous especially at large topology scopes, until we have some mechanism like priority/preemption. It could be used maliciously or by accident to block scheduling in an entire cluster, for example. The primary use case for hard anti-affinity is at the node level (give a pod exclusive access to a machine); we don't anticipate that larger topology scopes are urgently needed. So before deploying anti-affinity, we should build an admission controller that is enabled by default that prohibits anti-affinity with a TopologyKey other than "node."

cc/ @kevin-wangzefeng @bgrant0607

### PR DESCRIPTION
**Pull request by kevin-wangzefeng**
Closes #19914, implements admission controller to restrict hard ("requiredDuringScheduling") anti-affinity to "node" TopologyKey

Please note that the first 10 commits is in another PR(#22985)

@davidopp PTAL.

### CODE REVIEW

File: plugin/pkg/admission/antiaffinity/admission.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+	"k8s.io/kubernetes/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	apierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+)
+
+func init() {
+	admission.RegisterPlugin("LimitPodAntiAffinityTopology", func(client clientset.Interface, config io.Reader) (admission.Interface, error) {
```
Comment (davidopp) [inline plugin/pkg/admission/antiaffinity/admission.go]: Better name: LimitPodHardAntiAffinityTopology

File: plugin/pkg/admission/antiaffinity/admission.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		if len(affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution) != 0 {
+			podAntiAffinityTerms = affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution
+		}
+		// TODO: Uncomment this block when implement RequiredDuringSchedulingRequiredDuringExecution.
+		//if len(affinity.PodAntiAffinity.RequiredDuringSchedulingRequiredDuringExecution) != 0 {
+		//        podAntiAffinityTerms = append(podAntiAffinityTerms, affinity.PodAntiAffinity.RequiredDuringSchedulingRequiredDuringExecution...)
+		//}
+		for _, v := range podAntiAffinityTerms {
+			if v.TopologyKey != unversioned.LabelHostname {
+				return apierrors.NewForbidden(a.GetResource(), pod.Name, fmt.Errorf("affinity.PodAntiAffinity has TopologyKey other than :%v is forbidden", unversioned.LabelHostname))
```
Comment (davidopp) [inline plugin/pkg/admission/antiaffinity/admission.go]: Maybe include the specified TopologyKey in the error message, e.g.
```
fmt.Errorf("affinity.PodAntiAffinity.RequiredDuringScheduling has TopologyKey %v but only key %v is allowed", v.TopologyKey, unversioned.LabelHostname)
```

File: plugin/pkg/admission/antiaffinity/admission_test.go [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+package antiaffinity
+
+import (
+	"k8s.io/kubernetes/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"testing"
+)
+
+// ensures the PodAntiAffinity is denied if it defines topologykey anything other than kubernetes.io/hostname.
```
Comment (davidopp) [inline plugin/pkg/admission/antiaffinity/admission_test.go]: s/PodAntiAffinity/hard PodAntiAffinity/
s/topologykey anything/TopologyKey/

File: plugin/pkg/admission/antiaffinity/admission_test.go [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+								"namespaces":[{}],
+								"topologyKey": "az"
+							}
+						}]
+					}}`,
+			},
+			errorExpected: false,
+		},
+		{
+			affinity: map[string]string{
```
Comment (davidopp) [inline plugin/pkg/admission/antiaffinity/admission_test.go]: add a comment for this one, like
```
//  valid topologyKey in requiredDuringSchedulingIgnoredDuringExecution, plus any topologyKey in preferredDuringSchedulingIgnoredDuringExecution, then admission success.
```

**Comment by davidopp** [comment]
I reviewed the commit with the anti-affinity admission controller. Looks great, just had a couple of very minor comments.

One other suggestion: Can you add a `plugin/pkg/admission/antiaffinity/doc.go`
file to describe the admission controller? Here is a possible description:
"LimitPodHardAntiAffinityTopology admission controller rejects any pod that specifies "hard" (RequiredDuringScheduling) anti-affinity with a TopologyKey other than unversioned.LabelHostname. Because anti-affinity is symmetric, without this admission controller, a user could maliciously or accidentally specify that their pod (once it has scheduled) should block other pods from scheduling into the same zone or some other large topology, essentially DoSing the cluster. In the future we will address this problem more fully by using quota and priority, but for now this admission controller provides a simple protection, on the assumption that the only legitimate use of hard pod anti-affinity is to exclude other pods from the same node."

BTW, just to verify, this controller is disabled by default?

**Comment by kevin-wangzefeng** [comment]
@davidopp refreshed, PTAL

**Comment by davidopp** [comment]
LGTM

Thanks!

**Comment by davidopp** [comment]
Ah, I need to manually merge this due to the change to docs/admin/kube-apiserver.md

**Comment by kevin-wangzefeng** [comment]
@davidopp Thanks!
One thing left is: we need to make it possible to test admission controllers not enabled by default in e2e. Then I can add e2e test cases for this admission controller even it's not enabled by default.

**Pull request merged**